<a href="https://colab.research.google.com/github/somas1/CT/blob/main/NBA_PANDAS/NBA_Regular_Season_2018_19_Data_Challenge.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Nba Regular Season 2018-19 Data Challenge

Your task will be to take the dataset given, and create an analysis answering the following 10 questions. This project will again test your knowledge of pandas in order to find the answers needed given the data you are presented with.

In [2]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

In [3]:
nba = pd.read_csv('nbastats2018-2019.csv')
nba.head()

Name  Height  Weight                   Team  Age    Salary  Points  \
0  Alex Abrines      78     200  Oklahoma City Thunder   25   5455236     5.3   
1    Quincy Acy      79     240           Phoenix Suns   28    213949     1.7   
2  Jaylen Adams      74     190          Atlanta Hawks   22    236854     3.2   
3  Steven Adams      84     265  Oklahoma City Thunder   25  24157304    13.9   
4   Bam Adebayo      82     255             Miami Heat   21   2955840     8.9   

   Blocks  Steals  Assists  ...    MP   G   PER  OWS  DWS   WS   WS48   USG  \
0     0.2     0.5      0.6  ...  19.0  31   6.3  0.1  0.6  0.6  0.053  12.2   
1     0.4     0.1      0.8  ...  12.3  10   2.9 -0.1  0.0 -0.1 -0.022   9.2   
2     0.1     0.4      1.9  ...  12.6  34   7.6 -0.1  0.2  0.1  0.011  13.5   
3     1.0     1.5      1.6  ...  33.4  80  18.5  5.1  4.0  9.1  0.163  16.4   
4     0.8     0.9      2.2  ...  23.3  82  17.9  3.4  3.4  6.8  0.171  15.8   

   BPM  VORP  
0 -3.4  -0.2  
1 -5.9  -0.1  
2 -4.4  -0.3  
3  2.7   3.2  
4  3.0   2.4  

[5 rows x 27 columns]

# What was the average age of player in the league?

In [152]:
nba['Age'].mean().round(2)

25.9

# What player scored the most points?

In [39]:
nba.sort_values('Points', ascending=False)[['Name', 'Points']].head(1)

Name  Points
202  James Harden    36.1

James Harden had a higher point value than any other player in the dataset at 36.1. This suggests the values in the dataset are averages for the season, not totals.

Since the values presented are most likely averages, I will avoid applying functions such as .mean() to the dataset to average values further.

# What player had the most blocks during the season? Was it a post player (F/C)?

In [136]:
most_blocks = nba.sort_values(by='Blocks', ascending=False)[['Name', 'Height', 'Points', 'Blocks', 'Rebounds', 'Assists', 'FGA']].head(1)
most_blocks

Name  Height  Points  Blocks  Rebounds  Assists   FGA
476  Myles Turner      83    13.3     2.7       7.2      1.6  10.5

Myles Turner had the most blocks during the 2018-2019 season. He is 83 inches or 6'9 inches tall which suggests he might play center. Center players are usually post players.

A post player plays with their back to the basket, which gives them the ability to block more shots than they can typically make.

Myles has averaged 7.6 rebounds and 1.2 assists a game. He's attempted 10.2 non-free throw shots. All of this suggests he's a post player.

# Based on the regular season, who had the best chance to win a title given their win percentage?

With the information available in the dataset, we will have to use the WS value or Win Shares to estimate how many wins a team had for the 2018-1019 season.

In [99]:
nba.groupby('Team').sum()

<ipython-input-99-271e6a9c6b0e>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  nba.groupby('Team').sum()


Height  Weight  Age  Points  Blocks  Steals  Assists  \
Team                                                                           
Atlanta Hawks             1667    4580  556   179.2     7.7    12.2     36.9   
Boston Celtics            1178    3391  384   136.4     7.2    10.4     31.9   
Brooklyn Nets             1422    3892  460   165.7     5.9     9.8     32.4   
Charlotte Hornets         1103    3031  364   135.3     6.1     8.6     28.8   
Chicago Bulls             1261    3461  389   159.8     6.9    10.6     31.5   
Cleveland Cavaliers       1269    3476  416   133.5     4.0     9.1     28.0   
Dallas Mavericks          1186    3193  418   131.9     5.0     8.3     30.3   
Denver Nuggets            1182    3221  379   137.5     5.5     9.6     33.0   
Detroit Pistons           1181    3321  394   119.2     5.0     7.7     25.7   
Golden State Warriors     1198    3336  418   145.9     9.6    10.1     37.9   
Houston Rockets           1025    2870  363   123.2     5.5     9.7     24.7   
Indiana Pacers            1339    3653  426   136.9     6.7    10.8     33.5   
Los Angeles Clippers      1488    3970  515   152.5     6.9    10.0     33.1   
Los Angeles Lakers        1263    3458  426   155.5     6.8    11.1     38.8   
Memphis Grizzlies         1336    3591  436   147.2     7.5    12.2     32.9   
Miami Heat                 873    2485  310   118.4     6.0     8.1     28.3   
Milwaukee Bucks           1103    3139  377   156.0     6.9     9.5     31.9   
Minnesota Timberwolves    1257    3490  426   157.8     7.0    12.3     38.2   
New Orleans Pelicans      1419    3933  464   174.5     8.2    12.0     42.1   
New York Knicks           1279    3610  394   149.4     8.5    10.8     31.5   
Oklahoma City Thunder      943    2545  309   126.6     5.8    10.4     26.4   
Orlando Magic             1183    3195  368   131.9     6.7     8.9     33.5   
Philadelphia 76ers        1362    3750  448   163.0     7.3    11.1     36.2   
Phoenix Suns              1658    4466  520   176.2     8.5    16.2     36.9   
Portland Trail Blazers    1116    3041  348   135.0     5.9     7.9     26.5   
Sacramento Kings          1186    3284  366   145.3     5.8     9.6     30.2   
San Antonio Spurs         1194    3297  422   126.9     5.8     7.2     28.6   
Toronto Raptors            944    2602  326   132.1     6.8     9.6     30.7   
Utah Jazz                 1184    3324  407   135.6     7.2    10.2     31.3   
Washington Wizards        1349    3802  455   164.3     6.3    12.2     38.1   

                        Rebounds     FT%   FTA  ...    PER   OWS   DWS    WS  \
Team                                            ...                            
Atlanta Hawks               73.2  14.091  36.1  ...  285.0  15.2  12.3  28.0   
Boston Celtics              54.3  11.822  24.0  ...  234.9  26.1  26.0  52.0   
Brooklyn Nets               71.0  11.682  38.5  ...  245.6  22.7  22.9  45.4   
Charlotte Hornets           55.5  10.861  29.1  ...  207.8  23.8  14.9  38.8   
Chicago Bulls               64.9  12.054  33.2  ...  216.2   9.0  12.0  20.9   
Cleveland Cavaliers         65.0  11.477  27.0  ...  196.0  11.9   4.1  16.2   
Dallas Mavericks            49.8  10.820  29.2  ...  206.6  18.0  15.6  33.7   
Denver Nuggets              55.5   9.904  26.3  ...  196.7  27.3  23.2  50.5   
Detroit Pistons             53.2  11.200  26.7  ...  180.2  17.8  21.1  38.9   
Golden State Warriors       62.7  10.421  26.6  ...  224.5  34.1  21.8  55.9   
Houston Rockets             47.4   7.539  25.7  ...  220.0  29.3  16.4  45.7   
Indiana Pacers              55.8  11.702  27.5  ...  206.3  19.0  28.5  47.7   
Los Angeles Clippers        64.0  12.939  35.4  ...  223.2  26.3  20.6  46.5   
Los Angeles Lakers          60.4  10.828  31.2  ...  194.9  14.2  20.5  34.8   
Memphis Grizzlies           56.8  13.576  31.1  ...  206.7  11.4  19.5  30.9   
Miami Heat                  51.9   7.556  24.5  ...  152.2  13.0  22.8  35.7   
Milwaukee Bucks        

# What player had the best 3-pt percentage?

In [160]:
FG3_percentage = nba.sort_values('FG3%', ascending=False)
FG3_percentage[['Name','FG3%','FG3A']].head(5)

Name   FG3%  FG3A
312  Scott Machado  1.000   0.3
439  Jordan Sibert  1.000   1.0
147   Trevon Duval  1.000   0.3
352  Eric Moreland  1.000   0.2
95    Troy Caupain  0.667   0.8

Scott Machado, Jordan Sibert, Trevon Duval and Eric Moreland all have a 100% 3 point attempt percentage. Jordan Sibert has attempted on average one 3 point attempt per game as well.

# Who played the most minutes during the season

In [5]:
nba['Most Minutes'] = nba['MP']*nba['G']
nba.sort_values(by='Most Minutes', ascending=False)[['Name', 'G', 'MP', 'Most Minutes']].head(1)

Name   G    MP  Most Minutes
37  Bradley Beal  82  36.9        3025.8

# What player given their player effiecency rating was the clutchest during the season?

In [11]:
nba.groupby('PER').max().sort_values(by='PER', ascending=False)[['Name','G', 'Most Minutes']].head(10)

<ipython-input-11-2d24d5efc253>:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.max is deprecated. In a future version, a TypeError will be raised. Before calling .max, select only columns which should be valid for the function.
  nba.groupby('PER').max().sort_values(by='PER', ascending=False)[['Name','G', 'Most Minutes']].head(10)


Name   G  Most Minutes
PER                                          
80.4                Zhou Qi   1           1.0
38.3           Trevon Duval   3           6.0
36.9         Gary Payton II   3          15.9
32.9          Alan Williams   5          26.0
30.9  Giannis Antetokounmpo  72        2361.6
30.7           Troy Caupain   4          16.0
30.6           James Harden  78        2870.4
30.3          Anthony Davis  56        1848.0
29.7          Jordan Sibert   1           4.0
26.3           Nikola Jokic  80        2548.7

Zhou Qi had the highest 'PER' score during the season but he only played for one game and only for a minute.

Giannis Anetetokounmpo and James Harden both played over 70 games and mainted a PER score over 30. They are the most consistent performers during the season.

# What team had the youngest roster?

In [32]:
youngest = nba.groupby('Team').mean()['Age'].sort_values().head()
youngest

<ipython-input-32-155512425cc1>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  youngest = nba.groupby('Team').mean()['Age'].sort_values().head()


Team
Chicago Bulls       24.312500
Sacramento Kings    24.400000
Orlando Magic       24.533333
New York Knicks     24.625000
Phoenix Suns        24.761905
Name: Age, dtype: float64

The Chicago Bulls had the team with the youngest mean age.

# Who is the highest paid player during the seasion?

In [60]:
nba[['Salary']].dtypes

Salary    object
dtype: object

The Salary column has strings rather than numeric values. This requires converting these strings to numbers before they can be sorted.

In [63]:
nba['Fixed-Salary'] = pd.to_numeric(nba['Salary'], errors='coerce')
nba.dtypes[['Salary','Fixed-Salary']]

Salary           object
Fixed-Salary    float64
dtype: object

In [70]:
nba[['Salary','Fixed-Salary']]

Salary  Fixed-Salary
80         -           NaN
314        -           NaN
315        -           NaN
192        -           NaN
322        -           NaN
..       ...           ...
157  9600000     9600000.0
129  9607500     9607500.0
402  9631250     9631250.0
188   988464      988464.0
255    99290       99290.0

[521 rows x 2 columns]

Null values are now NaN in the new Fixed-Salary column.

In [80]:
fixed_nba = nba.dropna()
fixed_nba[['Salary','Fixed-Salary']]

Salary  Fixed-Salary
313   1000000     1000000.0
57    1000000     1000000.0
193  10000000    10000000.0
107  10000000    10000000.0
297  10002681    10002681.0
..        ...           ...
157   9600000     9600000.0
129   9607500     9607500.0
402   9631250     9631250.0
188    988464      988464.0
255     99290       99290.0

[434 rows x 2 columns]

In [85]:
fixed_nba.sort_values(by='Fixed-Salary', ascending=False)
fixed_nba[['Name','Fixed-Salary']].head(1)


Name  Fixed-Salary
121  Stephen Curry    37457154.0

Stephen Curry was the highest paid athlete in the dataset.

# At the end of a game, who WOULDN'T you want on the Free Throw Line?

We will use fixed_nba so that NaN values are removed. This removes the players in the dataset who are not associated with any teams.

In [93]:
fixed_nba.sort_values(by='FT%')[['Name','FT%','FTA','Team']].head(10)

Name    FT%  FTA                    Team
154       Jacob Evans  0.000  0.0   Golden State Warriors
360       Dzanan Musa  0.000  0.2           Brooklyn Nets
268    Terrence Jones  0.000  0.5         Houston Rockets
325    Tahjere McCall  0.000  1.0           Brooklyn Nets
168    Melvin Frazier  0.250  0.4           Orlando Magic
310       Tyler Lydon  0.333  0.1          Denver Nuggets
451      Ray Spalding  0.333  0.9        Dallas Mavericks
324  Luc Mbah a Moute  0.400  1.3    Los Angeles Clippers
27         Lonzo Ball  0.417  1.0      Los Angeles Lakers
472    Gary Trent Jr.  0.429  0.5  Portland Trail Blazers

All of the players listed above have poor free throw attempts made during the 2018-2019 season but it looks like Tahjere McCall made an average of 1 free throw attempt per game with 0 shots made. He's the last person I'd want attempting a free throw.